libsel4camkes/docs/virtqueue.md

```diff
@@ -0,0 +1,133 @@
+can also be changed, however, it must be page-aligned to 4096.
+
+Next, in the application code, the driver and device side of the channel must
+initialise their end of the channel by calling the appropriate function from
+the following:
+
+```c
+static inline int camkes_virtqueue_driver_init(virtqueue_driver_t *driver, unsigned int camkes_virtqueue_id) { ... }
+
+static inline int camkes_virtqueue_device_init(virtqueue_device_t *device, unsigned int camkes_virtqueue_id) { ... }
+```
+
+There is also a version of these functions which can return the underlying
+notification object that is backing the notification part of the virtqueues.
+
+```c
+int camkes_virtqueue_driver_init_with_recv(virtqueue_driver_t *driver, unsigned int camkes_virtqueue_id, 
+                                           seL4_CPtr *recv_notification, seL4_CPtr *recv_badge);
+                                           
+int camkes_virtqueue_device_init_with_recv(virtqueue_device_t *device, unsigned int camkes_virtqueue_id,
+                                           seL4_CPtr *recv_notification, seL4_CPtr *recv_badge);
+```
+
+The rest of the functions are essentially helper or convenience functions over
+the virtqueue interface as defined
+[here](https://github.com/SEL4PROJ/projects_libs/blob/master/libvirtqueue/include/virtqueue.h).
+
+### Driver
+
+When sending data over to the device side, it is preferred to use to these two functions.
+
+```c
+int camkes_virtqueue_driver_send_buffer(virtqueue_driver_t *vq, void *buffer, size_t size);
+
+int camkes_virtqueue_driver_scatter_send_buffer(virtqueue_driver_t *vq, void *buffer, size_t size);
+```
+
+When collecting buffers from the device side, the client must first obtain a
+handle to a used buffer ring via this virtqueue interface function:
+
+```c
+int virtqueue_get_used_buf(virtqueue_driver_t *vq, virtqueue_ring_object_t *robj, uint32_t *len);
+```
+
+It is then possible to gather each buffer in the used buffer ring and copy it into a larger buffer uisng the first function or iterate through each buffer in the ring using the second function:
+
+```c
+int camkes_virtqueue_driver_gather_copy_buffer(virtqueue_driver_t *vq, virtqueue_ring_object *handle,
+                                        void *buffer, size_t size);
+                                        
+int camkes_virtqueue_driver_gather_buffer(virtqueue_driver_t *vq, virtqueue_ring_object_t *handle,
+                                          void **buffer, unsigned *size, vq_flags_t *flag);
+```
+
+### Device side
+
+Collecting buffers from the driver side requires the client to first obtain a
+handle to an available buffer ring via this virtqueue interface function:
+
+```c
+int virtqueue_get_available_buf(virtqueue_device_t *vq, virtqueue_ring_object_t *robj);
+```
+
+The buffers can then be:
+    - gathered and copied into a larger buffer using the first function,
+    - have the contents of a larger scatter-copied into the buffers using the second function,
```

Comment:
- For me - in VS Code with 'Markdown All in One' extension - the list is not rendered as a list. Is this working for you?
- The line width of 80 columns is exceeded for the 2nd bullet point.

Reply:
Yes, I think this should change. Ideally there should be a blank line before and after bullet lists in markdown.

(I tend to use markdownlint in VSCode, which is a bit strict, but so far has gotten me fairly portable results.)

This is still marked as draft, though, so we should give people some time :-)

Reply:
For me the problem was not missing blank lines but the indentation of the list.

(I am not very familiar with the process (yet), so please take your time ;-)

Reply:
Ah, yes, if it is indented by 4 it is probably interpreted as a code block.

Reply:
This than seems to depend on the blank line... The current "list" is just rendered as a text paragraph. This can also bee seen in the "Rich Diff" functionality in GitHub that shows the rendered text.